### DISCUSSION TITLE
Workspace naming is confusing

### DISCUSSION THREAD

**Discussion by jasonwilliams** [npm/feedback#510] [Documentation] [answered]
I may be missing something obvious, but why aren't packages within a workspace defined as such? There are contradictions among the documentation which interchanges between _packages_ and _workspaces_.
RFC: https://github.com/npm/rfcs/blob/latest/implemented/0026-workspaces.md

First off for me a workspace is the top-level folder which holds all of the packages that you _work_ in. But the inner packages which can be published to npm are _packages_.
The community seems to settle on this also: _A workspace is what npm calls a monorepo: A directory with subdirectories that are npm packages._ https://2ality.com/2021/07/simple-monorepos.html#npm-workspaces

In [dictionary](https://github.com/npm/rfcs/blob/latest/implemented/0026-workspaces.md#dictionary) we have a definition:

> **workspace**: A nested package within the Top-level workspace file system that is explicitly defined as such via workspaces configuration.
> **Top-level workspace**: The root level package that contains a workspaces configuration defining workspaces.

So my understanding here is top-level workspace is a package that containers configuration defining more (inner) workspaces. So it sounds like the inner packages are also called workspaces?

But the example lists them as packages.
```json
{
    "name": "workspace-example",
    "version": "1.0.0",
    "workspaces": {
        "packages": [
            "packages/*"
        ]
    }
}
```

So are they packages or workspaces?
The summary of the RFC calls them packages also:
> Add a set of features to the npm cli that provide support to managing multiple **packages** from within a singular top-level, root package.

and the motivation
> The primary motivation behind this RFC is to fully realize a set of features/functionality to support managing multiple **packages** that may or may not be used together.

**Notes**
https://github.com/npm/rfcs/pull/103

**Comment by arcanis** [comment]
> First off for me a workspace is the top-level folder which holds all of the packages that you work in. But the inner packages which can be published to npm are packages.

Fwiw that's not the case in Yarn, which is perhaps the main way the JS ecosystem is exposed to workspaces. To quote the nomenclature from our [documentation](https://yarnpkg.com/features/workspaces):

> First, some vocabulary: in the context of the workspace feature, a project is the whole directory tree making up your workspaces (often the repository itself). A workspace is a specific named package stored anywhere within the project. Finally, a worktree is the name given to packages that list their own child workspaces. A project contains one or more worktrees, which may themselves contain any number of workspaces.

I would discourage you from deviating from this definition if you intend to keep the `workspaces` property to define your workspaces, as it would be needlessly confusing for everyone. If you want to change it, you can do what pnpm did and use a different definition syntax.

> The community seems to settle on this also: _A workspace is what npm calls a monorepo: A directory with subdirectories that are npm packages._ https://2ality.com/2021/07/simple-monorepos.html#npm-workspaces

This is one article, versus years of consistent usage. Articles can make mistakes too.

**Reply by jasonwilliams** [reply]
@arcanis thanks but I don’t think the yarn definition is that much clearer, in fact I’m sure this rfc is based heavily off it.

For instance:
> A workspace is a specific named package stored anywhere within the project

That just sounds like a package. What’s the difference in this case? The definition is far too vague as it doesn’t separate a package and a workspace.

**Reply by arcanis** [reply]
> That just sounds like a package. What’s the difference in this case? The definition is far too vague as it doesn’t separate a package and a workspace.

A package is anything that contains a package.json from the dependency tree. A workspace is a package, but a package isn't necessarily a workspace (for instance, packages downloaded from the registry obviously aren't workspaces 😉).

**Reply by jasonwilliams** [reply]
> > That just sounds like a package. What’s the difference in this case? The definition is far too vague as it doesn’t separate a package and a workspace.
> 
> A package is anything that contains a package.json from the dependency tree. A workspace is a package, but a package isn't necessarily a workspace (for instance, packages downloaded from the registry obviously aren't workspaces 😉).

Sure but *your* definition doesn’t state that. It states a yarn workspace is a specifically named package. What’s the specific naming? Aren’t all packages named? What name should it have to become a workspace? It’s not a great definition by any standards and it’s too vague

**Reply by arcanis** [reply]
Right, I feel like those questions tend to go off topic - your initial statement was that workspaces are monorepos. My answer is that you're mistaken (in the context of Yarn, and npm's workspaces seem to follow the Yarn syntax), and I think the documentation is clear enough to support that.

**Reply by jasonwilliams** [reply]
> Right, I feel like those questions tend to go off topic 

It's not off-topic its trying to clarify the definition you set out. If those questions can't be answered then the definition breaks down.

> your initial statement was that workspaces are monorepos... I think the documentation is clear enough to support that.

That wasn't my initial statement, that was from the 2ality blogpost. The documentation isn't clear enough, otherwise there wouldn't be this thread. The fact that the definition you've given in this thread is different to the one on the yarn webpage should be an indicator in itself.

**Reply by ljharb** [reply]
A workspace, colloquially, is the place in which one works on projects, like a desk or a table. It makes no sense - for npm or for yarn - that the term “workspace” applies to the child projects within the overarching repo.

In lerna workspace discussions, which long predate yarn’s implementation, the term “workspace” always means the containing repo, not the sub projects.

**Reply by arcanis** [reply]
I'll remain out of the terminology opinions (I didn't pick the initial wording, and I only care about breaking changes), but out of curiosity which discussion are you referring to? I don't find much before 2017, which is when workspaces got introduced.

**Reply by ljharb** [reply]
Oh sorry, to clarify I don't necessarily mean initial discussion - i mean the way the community talked about lerna monorepos.

Also by "predate" I was referring to common usage; i don't actually know release dates for any of the implementations.

Either way, the non-computer use of "workspace" is as I described, which I think is why npm and yarn's current terminology always seems confusing in my experience to people new to it.

**Reply by arcanis** [reply]
Honestly, and perhaps that's because I'm a non-native english speaker, I can see both ways working out - a workspace being either a place where you work on multiple projects, or a workspace being one space of a larger area (for example, that's what I find in the [dictionary.com definition](https://www.dictionary.com/browse/workspace)).

Ultimately I have no doubt any interpretation will have its supporters, but in the end that seems largely bikeshedding to me, and I don't think it's a good reason for breaking what's already a consistent terminology, even if you don't agree with it and we perhaps can find a handful of articles who misunderstood or ignored it (which would happen regardless of your definition).

**Reply by ljharb** [reply]
I think that keeping it the same in yarn makes sense for that reason - but npm's use of the terminology is new enough that it makes a ton of sense to disrupt the tiny number of people using the feature now, so as to reduce confusion for the millions that will be using it later.

**Reply by arcanis** [reply]
I don't mind that *if* you use a different syntax. If you use the same, people writing articles about our syntax will have no way to write them in a way that's understandable by both package manager users, leading to a largely avoidable confusion depending on whether the articles / tutorials got written by a Yarn or npm user.

Note that generally speaking I also think it's an horrible idea for your own users, even if my focus is on protecting ours.

**Reply by ljharb** [reply]
I'm quite sure that long term the syntax won't be the same, so that won't be an issue.

**Reply by jasonwilliams** [reply]
> A workspace, colloquially, is the place in which one works on projects, like a desk or a table. It makes no sense - for npm or for yarn - that the term “workspace” applies to the child projects within the overarching repo.
> 
> In lerna workspace discussions, which long predate yarn’s implementation, the term “workspace” always means the containing repo, not the sub projects.

This is the same understanding I have.

> I don't think it's a good reason for breaking what's already a consistent terminology.

I think this terminology is far from consistent, even for yarn users (who have had this feature longer) there's mixed usage of the terms. This isn't just because people are ignoring it but because some of the existing material isn't straight forward, it doesn't help by calling both the outer project and child projects a workspace.

As @ljharb mentioned I think this is an opportunity to get it right before its used by millions in future.

**Reply by MylesBorins** [reply]
@jasonwilliams with workspaces already shipped in npm 7 and it being included in Node.js 16, I think it might be hard to make a breaking change like this

With that said a good next step would be opening an issue over at the [RFC repository](https://github.com/npm/rfcs) and see what the CLI team has to say. Perhaps there is a non-breaking way to accomplish this

**Reply by jasonwilliams** [reply]
@MylesBorins which bit do you see breaking? My request was just to change some of the terminology, how NPM functions on workspaces should remain the same.

**Reply by arcanis** [reply]
Breaking change is a thing in terms of terminology too. Quoting my previous comment:

> If you use the same, people writing articles about our syntax will have no way to write them in a way that's understandable by both package manager users, leading to a largely avoidable confusion depending on whether the articles / tutorials got written by a Yarn or npm user.

If I say that blue is now red, all articles / tutorials / support threads referring to red prior the change now don't make any sense.

**Reply by jasonwilliams** [reply]
Hey @arcanis thanks, I’m aware breaking changes should include documentation (I write a lot of APIs and docs myself). 
However here I’m referring to the RFC itself in response to @MylesBorins statement that “this has shipped”, no new code needs to be shipped in a new version to accommodate these changes.

> If I say that blue is now red, all articles / tutorials / support threads referring to red prior the change now don't make any sense.

Well as both me and @ljharb pointed out, this is a new feature, it’s better to address issues now before the problem gets worse for millions later. I don’t want to repeat everything that’s being said above but *not* doing anything will cause just as many problems (probably more) than doing something.

**Reply by ljharb** [reply]
This confusion comes up in literally every RFC call that discusses workspaces, as well as in almost every issue/PR about it. Sometimes when people say "workspace" they mean a child project, and sometimes they mean the repo itself. We've been dancing around it by using terms like "parent workspace" or "child workspace", but clarifying the terms will help not just in docs, but also so discussion about the feature uses more universally shared terminology.

**Reply by MylesBorins** [reply]
I may have misunderstood and assumed that this suggested change would have involved changes to the manifest / configuration, not only to definitions + documentation.

Either way I do think it is worth opening an RFC to discuss the larger change.

**Reply by isaacs** [reply]
Was going to reply here, but it got too long.  https://github.com/npm/feedback/discussions/510#discussioncomment-1155331

tl;dr - I agree with @arcanis and I think npm should continue trying to match yarn's definitions.

**Comment by isaacs** [comment, marked as answer]
Naming things is hard.  Human language is a slippery beast at best.  (This is similar to the module/package/dependency semantic rats' nest, and occasionally gets entangled with it.)

We have been _trying_ to converge on this terminology, which I believe is compatible with Yarn's semantics:

* `project` or `project root` - The parent directory where you run npm commands, which has a package.json, and is typically the root of the git repository.  This may include a `workspaces` definition in its package.json file.  This may be a published package, or just a parent of workspaces that does not define any other functionality of its own.  If it defines `workspaces` in package.json, then you may use npm commands with the `--workspace=<name>` or `--workspaces` options.
* `workspace` or `child workspace` - The thing that is referenced by `workspaces` in the `project`.
* `workspace root` - A `project root` that includes a `workspaces` definition.
* `package` - A thing that can be operated on by npm and referenced by a package specifier.  A tarball file or url, `name@spec`, a folder path containing a package.json, etc.
* `manifest` - A set of metadata describing a package, typically written in `package.json`, but also served by the registry, loaded from a lockfile, etc.
* `dependency`
    1. A stated requirement of a package, listed in `dependencies`, `devDependencies`, etc. of a manifest.
    2. A package satisfies a `dependency(1)`.  Typically, a folder in `node_modules`, but may also be an abstract representation of a package in a manifest.
* `packument` - A document describing all published versions of a package with a given name, including their dist-tags, manifests, and other metadata.
* `module` - A thing that may be loaded by the `require()` or `import` functions in JavaScript.  A `package` is typically a module if it includes a `main` field in its manifest, or if it contains an `index.js` file.
* `specifier` - A representation of a `dependency(1)`, which may include the name of the dependency, either as a string (such as `foo@1.x`) or a parsed representation of it.

Within npm's internals (particularly Arborist), we also have the following more technical definitions:

* `node` - An abstract representation of a package in a dependency graph, which may be reified to or loaded from either the file system or a lockfile.
* `link` - An abstract representation of a symbolic link to a `node`.
* `edge` - An abstract representation of a `dependency(1)`, including the `node` with the dependency, the type and `specifier` defining it, the `link` or `node` that it will resolve to (if any), and the error state if the dependency represented is not properly met.
* `workspace` - (more precisely than above) A `node` that is included in the `workspaces` definition of a project, or a `link` to a workspace node.
* `root` - The `node` representing the `project root`.

npm does not (yet?) have a concept of "worktrees".

I do not personally have any strong opinion on how these terms are defined, but I am very much in agreement with @arcanis that it's valuable to have npm and yarn be in sync here.  Since yarn implemented workspaces a few years ahead of npm, yarn has set the definitions, and that's fine.  Changing them now seems like it would not be worthwhile.

I am willing to bet that some of npm's documentation is currently using these terms imprecisely or even incorrectly.  Those cases are documentation bugs that should be fixed.

I do realize that lerna predates yarn, and yarn deviated from lerna's naming conventions.  But (a) npm is more like yarn than it is like lerna, and (b) we have to pick one to follow, and we picked yarn, and here we are.  Picking lerna's definitions instead would have also been fine, but it's not how it went.
